File: app/src/main/java/com/rachelleignacio/listbucket/presentation/presenters/impl/ListItemsFragmentPresenterImpl.java
package com.rachelleignacio.listbucket.presentation.presenters.impl;

import com.rachelleignacio.listbucket.db.DbInteractor;
import com.rachelleignacio.listbucket.domain.executor.MainThread;
import com.rachelleignacio.listbucket.domain.executor.ThreadExecutor;
import com.rachelleignacio.listbucket.domain.interactors.AddListItemInteractor;
import com.rachelleignacio.listbucket.domain.interactors.DeleteListItemInteractor;
import com.rachelleignacio.listbucket.domain.interactors.GetAllListItemsInteractor;
import com.rachelleignacio.listbucket.domain.interactors.impl.AddListItemInteractorImpl;
import com.rachelleignacio.listbucket.domain.interactors.impl.DeleteListItemInteractorImpl;
import com.rachelleignacio.listbucket.domain.interactors.impl.GetAllListItemsInteractorImpl;
import com.rachelleignacio.listbucket.domain.models.ListItem;
import com.rachelleignacio.listbucket.presentation.presenters.AbstractPresenter;
import com.rachelleignacio.listbucket.presentation.presenters.ListItemsFragmentPresenter;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by rachelleignacio on 3/31/17.
 */

public class ListItemsFragmentPresenterImpl extends AbstractPresenter implements
        ListItemsFragmentPresenter, GetAllListItemsInteractor.Callback,
        AddListItemInteractor.Callback, DeleteListItemInteractor.Callback {

    private DbInteractor dbInteractor;
    private ListItemsFragmentPresenter.View view;
    private com.rachelleignacio.listbucket.domain.models.List parentList;
    private java.util.List<ListItem> listItems;

    public ListItemsFragmentPresenterImpl(ThreadExecutor threadExecutor, MainThread mainThread, DbInteractor db,
                                          View view, com.rachelleignacio.listbucket.domain.models.List parent) {
        super(threadExecutor, mainThread);
        this.dbInteractor = db;
        this.view = view;
        this.parentList = parent;
    }

    @Override
    public void getListItems() {
        GetAllListItemsInteractor getItemsInteractor = new GetAllListItemsInteractorImpl(threadExecutor,
                mainThread, this, dbInteractor, parentList.getId());
        getItemsInteractor.execute();
    }

    @Override
    public void onListItemsRetrieved(List<ListItem> items) {
        listItems = new ArrayList<>();
        listItems.addAll(items);
        view.showListItems(listItems);
    }

    @Override
    public void addListItem(String listItemName) {
        ListItem itemToSave = new ListItem(parentList, listItemName);
        AddListItemInteractor addItemInteractor = new AddListItemInteractorImpl(threadExecutor, mainThread,
                this, dbInteractor, itemToSave);
        addItemInteractor.execute();
    }

    @Override
    public void onListItemAdded(ListItem newItem) {
        listItems.add(newItem);
        view.showListItems(listItems);
    }

    @Override
    public void deleteListItem(int position) {
        DeleteListItemInteractor deleteListItemInteractor = new DeleteListItemInteractorImpl(threadExecutor,
                mainThread, this, dbInteractor, listItems.get(position), position);
        deleteListItemInteractor.execute();
    }

    @Override
    public void onListItemDeleted(int position) {
        listItems.remove(position);
        view.showListItems(listItems);
    }

    @Override
    public int getListCount() {
        return listItems.size();
    }
}

File: app/src/main/kotlin/com/rachelleignacio/listbucket/presentation/presenters/impl/ListItemsFragmentPresenterImpl.kt
package com.rachelleignacio.listbucket.presentation.presenters.impl

import com.rachelleignacio.listbucket.db.DbInteractor
import com.rachelleignacio.listbucket.domain.executor.MainThread
import com.rachelleignacio.listbucket.domain.executor.ThreadExecutor
import com.rachelleignacio.listbucket.domain.interactors.AddListItemInteractor
import com.rachelleignacio.listbucket.domain.interactors.DeleteListItemInteractor
import com.rachelleignacio.listbucket.domain.interactors.GetAllListItemsInteractor
import com.rachelleignacio.listbucket.domain.interactors.impl.AddListItemInteractorImpl
import com.rachelleignacio.listbucket.domain.interactors.impl.DeleteListItemInteractorImpl
import com.rachelleignacio.listbucket.domain.interactors.impl.GetAllListItemsInteractorImpl
import com.rachelleignacio.listbucket.domain.models.List
import com.rachelleignacio.listbucket.domain.models.ListItem
import com.rachelleignacio.listbucket.presentation.presenters.AbstractPresenter
import com.rachelleignacio.listbucket.presentation.presenters.ListItemsFragmentPresenter

/**
 * Created by rachelleignacio on 8/23/17.
 */
class ListItemsFragmentPresenterImpl(threadExecutor: ThreadExecutor,
                                     mainThread: MainThread,
                                     private val dbInteractor: DbInteractor,
                                     private val view: ListItemsFragmentPresenter.View,
                                     private val parentList: List
) : AbstractPresenter(threadExecutor, mainThread),
    ListItemsFragmentPresenter,
    GetAllListItemsInteractor.Callback,
    AddListItemInteractor.Callback,
    DeleteListItemInteractor.Callback
{

    private var listItems = mutableListOf<ListItem>()

    override fun getListItems() {
        GetAllListItemsInteractorImpl(threadExecutor, mainThread, this, dbInteractor, parentList.id)
            .execute()
    }

    override fun onListItemsRetrieved(items: MutableList<ListItem>) {
        listItems.clear()
        listItems.addAll(items)
        view.showListItems(listItems)
    }

    override fun addListItem(listItemName: String) {
        val itemToSave = ListItem(parentList, listItemName)
        AddListItemInteractorImpl(threadExecutor, mainThread, this, dbInteractor, itemToSave)
                .execute()
    }

    override fun onListItemAdded(newItem: ListItem) {
        listItems.add(newItem)
        view.showListItems(listItems)
    }

    override fun deleteListItem(adapterPosition: Int) {
        DeleteListItemInteractorImpl(threadExecutor, mainThread, this, dbInteractor, listItems[adapterPosition], adapterPosition)
                .execute()
    }

    override fun onListItemDeleted(adapterPosition: Int) {
        listItems.removeAt(adapterPosition)
        view.showListItems(listItems)
    }

    override val listCount: Int
        get() = listItems.size
}